src/main/java/org/zalando/nakadi/service/ConsumerLimitingService.java

```diff
@@ -0,0 +1,231 @@
+        final boolean hasMoreConnectionsToPartition = ACQUIRED_SLOTS.stream()
+                .anyMatch(s -> s.getPartition().equals(slot.getPartition())
+                        && s.getClient().equals(slot.getClient())
+                        && s.getEventType().equals(slot.getEventType()));
+        if (!hasMoreConnectionsToPartition) {
+            final String consumerPath = zkPathForConsumer(slot.getClient(), slot.getEventType(), slot.getPartition());
+            final PathChildrenCache cache = SLOTS_CACHES.getOrDefault(consumerPath, null);
+            if (cache != null) {
+                SLOTS_CACHES.remove(consumerPath);
+                cache.close();
+            }
+        }
+    }
+
+    public void deletePartitionNodeIfPossible(final String nodeName) {
+        try {
+            zkHolder.get()
+                    .delete()
+                    .forPath(CONNECTIONS_ZK_PATH + "/" + nodeName);
+        } catch (final KeeperException.NotEmptyException e) {
+            // if the node has children - we should not delete it
+        } catch (final Exception e) {
+            LOG.error("Zookeeper error when trying delete consumer node", e);
+        }
+    }
+
+    private Optional<ConnectionSlot> acquireConnectionSlot(final String client, final String eventType,
+                                                           final String partition)
+            throws ConnectionSlotOccupiedException {
+
+        final String parent = zkPathForConsumer(client, eventType, partition);
+
+        final List<String> occupiedSlots = getChildrenCached(parent);
+        if (occupiedSlots.size() >= maxConnections) {
+            return Optional.empty();
+        }
+
+        final List<String> availableSlots = slotNames.stream()
+                .filter(slot -> !occupiedSlots.contains(slot))
+                .collect(toList());
+        final int slotIndex = new Random().nextInt(availableSlots.size());
+        final String slot = availableSlots.get(slotIndex);
+
+        final String zkPath = parent + "/" + slot;
+        try {
+            zkHolder.get()
+                    .create()
+                    .creatingParentsIfNeeded()
+                    .withMode(CreateMode.EPHEMERAL)
+                    .forPath(zkPath);
+        } catch (final KeeperException.NodeExistsException e) {
+            throw new ConnectionSlotOccupiedException();
+        } catch (Exception e) {
+            LOG.error("Zookeeper error when creating consumer node", e);
+            throw new NakadiRuntimeException(e);
+        }
+
+        final ConnectionSlot acquiredSlot = new ConnectionSlot(client, eventType, partition, slot);
+        ACQUIRED_SLOTS.add(acquiredSlot);
+        return Optional.of(acquiredSlot);
+    }
+
+    private List<String> getChildrenCached(final String zkPath) {
+        try {
+            PathChildrenCache cache = SLOTS_CACHES.getOrDefault(zkPath, null);
```

Comment:
I don't really need to cache all the nodes' children in "/nakadi/consumers/connections"
I'm only interested in nodes's children for connections that exist in this Nakadi instance. 
So the TreeCache is too much for me in this case, it will create a huge amount of watchers which I really don't need and that will put a very high load on ZK.